lib/net.js

```diff
@@ -1333,58 +1344,73 @@ function listen(self, address, port, addressType, backlog, fd, exclusive) {
-  var backlog = toNumber(args.length > 1 && args[1]) ||
-                toNumber(args.length > 2 && args[2]);
-
-  options = options._handle || options.handle || options;
+  // (handle[, backlog][, cb]) where handle is an object with a handle
+  const handle = options._handle || options.handle;
+  if (handle && handle instanceof TCP) {
+    this._handle = handle;
+    listen(this, null, -1, -1, backlogFromArgs);
+    return this;
+  }
+  // (handle[, backlog][, cb]) where handle is an object with a fd
+  if (typeof options.fd === 'number' && options.fd >= 0) {
+    listen(this, null, null, null, backlogFromArgs, options.fd);
+    return this;
+  }
 
-  if (options instanceof TCP) {
-    this._handle = options;
-    listen(this, null, -1, -1, backlog);
-  } else if (typeof options.fd === 'number' && options.fd >= 0) {
-    listen(this, null, null, null, backlog, options.fd);
-  } else {
-    backlog = options.backlog || backlog;
-
-    if (typeof options.port === 'number' || typeof options.port === 'string' ||
-        (typeof options.port === 'undefined' && 'port' in options)) {
-      // Undefined is interpreted as zero (random port) for consistency
-      // with net.connect().
-      assertPort(options.port);
-      if (options.host) {
-        lookupAndListen(this, options.port | 0, options.host, backlog,
-                        options.exclusive);
-      } else {
-        listen(this, null, options.port | 0, 4, backlog, undefined,
-               options.exclusive);
-      }
-    } else if (options.path && isPipeName(options.path)) {
-      // UNIX socket or Windows pipe.
-      const pipeName = this._pipeName = options.path;
-      listen(this, pipeName, -1, -1, backlog, undefined, options.exclusive);
-    } else {
-      throw new Error('Invalid listen argument: ' + options);
+  // ([port][, host][, backlog][, cb]) where port is omitted,
+  // that is, listen() or listen(cb),
+  // or (options[, cb]) where options.port is explicitly set as undefined,
+  // bind to an arbitrary unused port
+  if (args.length === 0 || typeof args[0] === 'function' ||
+    (typeof options.port === 'undefined' && 'port' in options)) {
+    options.port = 0;
+  }
+  // ([port][, host][, backlog][, cb]) where port is specified
+  // or (options[, cb]) where options.port is specified
+  // or if options.port is normalized as 0 before
+  if (typeof options.port === 'number' || typeof options.port === 'string') {
+    if (!isLegalPort(options.port)) {
+      throw new RangeError('"port" argument must be >= 0 and < 65536');
+    }
+    options.port = options.port | 0;
+    const backlog = options.backlog || backlogFromArgs;
+    // start TCP server listening on host:port
+    if (options.host) {
+      lookupAndListen(this, options.port, options.host, backlog,
+                      options.exclusive);
+    } else { // Undefined host, listens on unspecified IPv4 address
+      listen(this, null, options.port, 4, backlog, undefined,
```

Comment:
The [doc](https://nodejs.org/api/net.html#net_server_listen_port_hostname_backlog_callback) says

> If the hostname is omitted, the server will accept connections on the unspecified IPv6 address (::) when IPv6 is available, or the unspecified IPv4 address (0.0.0.0) otherwise.

But looks like it just went straight to IPv4?